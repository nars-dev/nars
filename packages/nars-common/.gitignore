node_modules/
tsconfig.tsbuildinfo
.merlin
lib
