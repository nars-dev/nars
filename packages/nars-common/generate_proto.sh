#!/bin/sh

yarn pbjs -t static-module -w commonjs -o ./dist/Schema.js ./src/proto/schema.proto
yarn pbts --no-comments -o ./src/Schema.d.ts ./dist/Schema.js
cp ./src/Schema.d.ts ./dist/Schema.d.ts

if [[ $1 == "--watch" ]]; then
  yarn watch ./generate_proto.sh ./src/proto
fi
