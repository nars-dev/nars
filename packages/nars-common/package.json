{
  "name": "nars-common",
  "version": "0.0.0",
  "description": "Package for nars code shared between client and server",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "repository": "https://github.com/nars-dev/nars",
  "author": "Wojtek Czekalski <wojtek@nars.dev>",
  "license": "MIT",
  "private": false,
  "scripts": {
    "build": "tsc",
    "watch": "tsc --watch"
  },
  "devDependencies": {
    "prettier": "^1.18.2",
    "typescript": "^3.5.3"
  }
}
