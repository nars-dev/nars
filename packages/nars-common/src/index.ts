export interface InputPropType<T> {
  decode: (obj: unknown) => T | undefined;
  encode: (obj: T) => unknown | undefined;
  optional?: boolean;
}

export type LocalPropRequired = "optional" | "required";

export interface LocalPropKey<
  Component,
  Key,
  IsRequired extends LocalPropRequired
> {
  component: Component;
  key: Key;
  local: true;
  isRequired: IsRequired;
}

export type PropType =
  | LocalPropKey<any, any, LocalPropRequired>
  | InputPropType<any>;

export type PropTypes = {
  [K: string]: PropType;
};

export type ComponentConfig = { [k: string]: PropTypes };

export type ExtractInputPropType<T> = T extends InputPropType<infer ParsedType>
  ? ParsedType
  : never;

export const InputProp = {
  string: {
    decode: (obj: unknown) => String(obj),
    encode: (obj: string) => obj,
  },
  optional: <T>({
    encode,
    decode,
  }: InputPropType<T>): InputPropType<T | undefined> => ({
    decode: (obj: unknown) => (obj ? decode(obj) : undefined),
    encode: (obj: T | undefined) => (obj ? encode(obj) : undefined),
    optional: true,
  }),
};

export function localProp<
  IsRequired extends LocalPropRequired,
  A extends string,
  B extends string
>(isRequired: IsRequired, a: A, b: B): LocalPropKey<A, B, IsRequired> {
  return { component: a, key: b, local: true, isRequired };
}
