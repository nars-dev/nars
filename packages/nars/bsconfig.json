{
  "name": "nars",
  "version": "0.0.0",
  "reason": {
    "react-jsx": 3
  },
  "sources": {
    "dir" : "src",
    "subdirs" : true
  },
  "package-specs": {
    "module": "commonjs",
    "in-source": true
  },
  "suffix": ".bs.js",
  "bs-dependencies": ["reason-react-reconciler", "ocaml-protoc-plugin-runtime-bs"],
  "warnings": {
    "error" : "+101"
  },
  "gentypeconfig": {
    "path": "./gentype.ppx",
    "module": "commonjs",
    "language": "typescript",
    "shims": {
      "Js": "Js",
      "ReactReconciler": "ReactReconciler"
    }
  },
  "refmt": 3
}
