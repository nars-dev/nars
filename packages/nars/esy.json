{
  "scripts": {
    "build-protoc": "protoc --ocaml_out=./src/protobuf/ --ocaml_opt=int64_as_int=false -I ../nars-common/proto/ nars_animated.proto struct.proto schema.proto"
  },
  "dependencies": { "ocaml": "4.6.x" },
  "devDependencies": {
    "@opam/dune": "1.11.4",
    "@opam/ocaml-lsp-server": "ocaml/ocaml-lsp:ocaml-lsp-server.opam#007a1af",
    "@opam/reason": "*",
    "@opam/ocamlformat": "~0.11.0",
    "@opam/ocaml-protoc-plugin": "^3.0.0"
  },
  "resolutions": {
    "@opam/reason": "facebook/reason#b555665"
  }
}
