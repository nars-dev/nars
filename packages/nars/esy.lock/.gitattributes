
# Set eol to LF so files aren't converted to CRLF-eol on Windows.
opam/** binary eol=lf linguist-generated=true
overrides/** binary eol=lf linguist-generated=true
