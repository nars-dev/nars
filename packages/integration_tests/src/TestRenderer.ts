import * as Client from "nars-client";
import WebSocket from "ws";
import {
  create,
  act,
  ReactTestRenderer,
  ReactTestInstance,
} from "react-test-renderer";
import { Static, Server, ComponentConfig } from "nars";

interface MessageEvent {
  readonly data: any;
}

// In reality we must realize that the code below should be tested in principle, too.
// We are not testing the real behavior. We are testing the behavior in this
// simulated enviornment which is vastly more predictable than the network conditions.
export type TestSocketLike = Client.SocketLike & {
  setServerSocketIn: (_: (data: Server.Socket_data) => any) => void;
  clientSocketIn: () => (ev: MessageEvent) => any;
};

export const testSocket: () => TestSocketLike = () => {
  let serverSocketIn: (data: Server.Socket_data) => any;
  let clientSocketIn: (ev: MessageEvent) => any;
  return {
    clientSocketIn: () => clientSocketIn,
    setServerSocketIn(x) {
      serverSocketIn = x;
    },
    readyState: WebSocket.OPEN,
    send(message) {
      if (serverSocketIn) {
        serverSocketIn((message as unknown) as Server.Socket_data);
      }
    },
    binaryType: "arraybuffer",
    addEventListener: (type, listener) => {
      if (type === "message") {
        clientSocketIn = listener;
      } else {
        throw "Unhandled event handler added in tests for type: " + type;
      }
    },
    removeEventListener: () => {},
  };
};

export const testSocketWithMessageSpy = (
  socket: TestSocketLike,
  eventHandlerSpy: (ev: MessageEvent) => any,
  messageSendSpy: (m: unknown) => void
): TestSocketLike => {
  const handler = socket.clientSocketIn;
  return {
    ...socket,
    clientSocketIn: () => ev => {
      eventHandlerSpy(ev);
      return handler()(ev);
    },
    send(d) {
      messageSendSpy(d);
      socket.send(d);
    },
  };
};

const server: (_: TestSocketLike) => Server.server = socket => ({
  on(event, handler) {
    if (event === "connection") {
      handler({
        on(event, handler) {
          if (event === "message") {
            socket.setServerSocketIn(handler);
          } else {
            throw "Unhandled event received: " + event;
          }
        },
        set binaryType(b) {
          socket.binaryType = b;
        },
        get binaryType() {
          return socket.binaryType;
        },
        send: msg => {
          if (socket.clientSocketIn()) {
            socket.clientSocketIn()({ data: msg });
          }
        },
      });
    }
  },
});

export const createRemoteComponentWithDefaultSocket = <
  T extends ComponentConfig
>(
  config: T,
  components: Static.ComponentDefinitions<T>
) => {
  const ws = testSocket();
  Static.attatchListener(server(ws), Static.createRouter(config, components));
  return Client.createRemoteComponentWithWebSocket(ws, config);
};

export const createRemoteComponent = <T extends ComponentConfig>(
  config: T,
  components: Static.ComponentDefinitions<T>
) => {
  return [
    Client.createRemoteComponent(config),
    (socket: TestSocketLike) =>
      Static.attatchListener(
        server(socket),
        Static.createRouter(config, components)
      ),
  ] as const;
};

export const render = (element: React.ReactElement): ReactTestRenderer => {
  let component: ReactTestRenderer;
  act(() => {
    component = create(element);
  });
  return component;
};

export const getChildren = (rendered: ReactTestRenderer) => {
  return (rendered.root.children[0] as ReactTestInstance).children;
};
