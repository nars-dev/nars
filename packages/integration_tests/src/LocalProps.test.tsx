jest.mock("react-native", () => ({
  FlatList: "FlatList",
  TouchableOpacity: "TouchableOpacity",
}));

import * as React from "react";
import { ReactTestInstance, ReactTestRenderer } from "react-test-renderer";
import { TouchableOpacity, FlatList, LocalProp } from "nars";
import { localProp } from "nars-common";
import { createRemoteComponent, render } from "./TestRenderer";

export const config = {
  TouchableOpacityTest: {
    props: { },
    localProps: {
      submit: localProp("TouchableOpacity", "onPress"),
    },
  },
  FlatListTest: {
    props: {},
    localProps: {
      reload: localProp("FlatList", "onEndReached"),
    },
  },
};

const components = {
  TouchableOpacityTest: (props: {
    props: { };
    localProps: { submit: LocalProp };
  }) => {
    return (
      <TouchableOpacity localProps={{ onPress: props.localProps.submit }} />
    );
  },
  FlatListTest: (props: { props: {}; localProps: { reload: LocalProp } }) => {
    return (
      <FlatList
        data={[]}
        keyExtractor={k => String(k)}
        renderItem={() => <TouchableOpacity />}
        localProps={{ onEndReached: props.localProps.reload }}
      />
    );
  },
};

const RemoteComponent = createRemoteComponent(config, components);

const getChildren = (rendered: ReactTestRenderer) => {
  return (rendered.root.children[0] as ReactTestInstance).children;
};

describe("Components", () => {
  it("TouhcableOpacity has onPress set to submit", async () => {
    const submit = () => null;
    const rendered = await render(
      <RemoteComponent
        name="TouchableOpacityTest"
        props={{
          props: { backgroundColor: "aa" },
          localProps: { submit },
        }}
      />
    );

    expect((getChildren(rendered)[0] as ReactTestInstance).props.onPress).toBe(
      submit
    );
  });
  it("FlatList has onEndReached set to reload", async () => {
    const reload = () => null;
    const rendered = await render(
      <RemoteComponent
        name="FlatListTest"
        props={{
          props: {},
          localProps: { reload },
        }}
      />
    );

    expect(
      (getChildren(rendered)[0] as ReactTestInstance).props.onEndReached
    ).toBe(reload);
  });
});
