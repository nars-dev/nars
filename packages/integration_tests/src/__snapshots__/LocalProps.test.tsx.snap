// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoteComponent Local Props works if an optional local prop is not passed in 1`] = `<TouchableOpacity />`;
